### PR TITLE
Fix crash on kube manager's service-lb-controller after v1.31.0.

### LINKED ISSUE #128121: Crash on kube manager's service-lb-controller after v1.31.0
### What happened?

If kube manager is started with no `cloud-provider`, it looks like the service-lb-controller is instantiated but not started. However, part of its initialization happens on the `Run` method **and** its handlers are added to the informer on instantiation. So the handlers still run and it crashes when they do because there is stuff not initialized, like in this case, the event recorder:
```2024-10-14T11:27:20.573127782Z stderr F 	goroutine 735 [running]:
2024-10-14T11:27:20.573132561Z stderr F 	k8s.io/apimachinery/pkg/util/runtime.logPanic({0x38432a0, 0x5545b00}, {0x2d99000, 0x5480790})
2024-10-14T11:27:20.573137951Z stderr F 		k8s.io/apimachinery/pkg/util/runtime/runtime.go:107 +0xbc
2024-10-14T11:27:20.573154772Z stderr F 	k8s.io/apimachinery/pkg/util/runtime.handleCrash({0x38432a0, 0x5545b00}, {0x2d99000, 0x5480790}, {0x5545b00, 0x0, 0x43d945?})
2024-10-14T11:27:20.573159501Z stderr F 		k8s.io/apimachinery/pkg/util/runtime/runtime.go:82 +0x5e
2024-10-14T11:27:20.573187343Z stderr F 	k8s.io/apimachinery/pkg/util/runtime.HandleCrash({0x0, 0x0, 0xc001e5ec40?})
2024-10-14T11:27:20.573192182Z stderr F 		k8s.io/apimachinery/pkg/util/runtime/runtime.go:59 +0x108
2024-10-14T11:27:20.57319655Z stderr F 	panic({0x2d99000?, 0x5480790?})
2024-10-14T11:27:20.573200548Z stderr F 		runtime/panic.go:770 +0x132
2024-10-14T11:27:20.573204836Z stderr F 	k8s.io/cloud-provider/controllers/service.(*Controller).needsUpdate(0xc000978000, 0xc00324c288, 0xc000cf3688)
2024-10-14T11:27:20.573210016Z stderr F 		k8s.io/cloud-provider/controllers/service/controller.go:606 +0xcbb
2024-10-14T11:27:20.573214644Z stderr F 	k8s.io/cloud-provider/controllers/service.New.func2({0x326c8e0?, 0xc00324c288?}, {0x326c8e0, 0xc000cf3688?})
2024-10-14T11:27:20.573219133Z stderr F 		k8s.io/cloud-provider/controllers/service/controller.go:144 +0x74
2024-10-14T11:27:20.573224062Z stderr F 	k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate(...)
2024-10-14T11:27:20.573228831Z stderr F 		k8s.io/client-go/tools/cache/controller.go:253
2024-10-14T11:27:20.573233339Z stderr F 	k8s.io/client-go/tools/cache.(*processorListener).run.func1()
2024-10-14T11:27:20.573237758Z stderr F 		k8s.io/client-go/tools/cache/shared_informer.go:976 +0xea
2024-10-14T11:27:20.573241875Z stderr F 	k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1(0x30?)
2024-10-14T11:27:20.57326562Z stderr F 		k8s.io/apimachinery/pkg/util/wait/backoff.go:226 +0x33
2024-10-14T11:27:20.573270419Z stderr F 	k8s.io/apimachinery/pkg/util/wait.BackoffUntil(0xc00311cf70, {0x380e700, 0xc001e3a8a0}, 0x1, 0xc001e3e780)
2024-10-14T11:27:20.573274767Z stderr F 		k8s.io/apimachinery/pkg/util/wait/backoff.go:227 +0xaf
2024-10-14T11:27:20.573279556Z stderr F 	k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc001e5a770, 0x3b9aca00, 0x0, 0x1, 0xc001e3e780)
2024-10-14T11:27:20.573284085Z stderr F 		k8s.io/apimachinery/pkg/util/wait/backoff.go:204 +0x7f
2024-10-14T11:27:20.573288433Z stderr F 	k8s.io/apimachinery/pkg/util/wait.Until(...)
2024-10-14T11:27:20.573292821Z stderr F 		k8s.io/apimachinery/pkg/util/wait/backoff.go:161
2024-10-14T11:27:20.573297059Z stderr F 	k8s.io/client-go/tools/cache.(*processorListener).run(0xc000892d80)
2024-10-14T11:27:20.573302339Z stderr F 		k8s.io/client-go/tools/cache/shared_informer.go:972 +0x69
2024-10-14T11:27:20.573306677Z stderr F 	k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
2024-10-14T11:27:20.573310975Z stderr F 		k8s.io/apimachinery/pkg/util/wait/wait.go:72 +0x52
2024-10-14T11:27:20.573315874Z stderr F 	created by k8s.io/apimachinery/pkg/util/wait.(*Group).Start in goroutine 690
2024-10-14T11:27:20.573320913Z stderr F 		k8s.io/apimachinery/pkg/util/wait/wait.go:70 +0x73
```

Might have been introduced with 
https://github.com/kubernetes/kubernetes/commit/50c12437604b0cd5a73514389409fc2fde8b91bd

### What did you expect to happen?

No crash.

### How can we reproduce it (as minimally and precisely as possible)?

```
❯ kind create cluster --name ovn
Creating cluster "ovn" ...
 ✓ Ensuring node image (kindest/node:v1.31.0) 🖼 
 ✓ Preparing nodes 📦  
 ✓ Writing configuration 📜 
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
Set kubectl context to "kind-ovn"
You can now use your cluster with:

kubectl cluster-info --context kind-ovn

Not sure what to do next? 😅  Check out https://kind.sigs.k8s.io/docs/user/quick-start/

❯ kind export kubeconfig --name ovn
Set kubectl context to "kind-ovn"

❯ kubectl get pods -A                                                                           
NAMESPACE            NAME                                        READY   STATUS    RESTARTS   AGE
kube-system          coredns-6f6b679f8f-2xrm5                    1/1     Running   0          2m57s
kube-system          coredns-6f6b679f8f-zpkp2                    1/1     Running   0          2m57s
kube-system          etcd-ovn-control-plane                      1/1     Running   0          3m6s
kube-system          kindnet-22kmz                               1/1     Running   0          2m58s
kube-system          kube-apiserver-ovn-control-plane            1/1     Running   0          3m5s
kube-system          kube-controller-manager-ovn-control-plane   1/1     Running   0          3m5s
kube-system          kube-proxy-8z75j                            1/1     Running   0          2m58s
kube-system          kube-scheduler-ovn-control-plane            1/1     Running   0          3m5s
local-path-storage   local-path-provisioner-57c5987fd4-sgvtl     1/1     Running   0          2m57s

❯ cat service.yaml                                                  
apiVersion: v1
kind: Service
metadata:
  name: example-service
spec:
  selector:
    app: example
  ports:
    - port: 8765
      targetPort: 9376
  type: LoadBalancer

❯ kubectl apply -f service.yaml    
service/example-service created

❯ kubectl get pods -A           
NAMESPACE            NAME                                        READY   STATUS    RESTARTS   AGE
kube-system          coredns-6f6b679f8f-2xrm5                    1/1     Running   0          5m15s
kube-system          coredns-6f6b679f8f-zpkp2                    1/1     Running   0          5m15s
kube-system          etcd-ovn-control-plane                      1/1     Running   0          5m24s
kube-system          kindnet-22kmz                               1/1     Running   0          5m16s
kube-system          kube-apiserver-ovn-control-plane            1/1     Running   0          5m23s
kube-system          kube-controller-manager-ovn-control-plane   1/1     Running   0          5m23s
kube-system          kube-proxy-8z75j                            1/1     Running   0          5m16s
kube-system          kube-scheduler-ovn-control-plane            1/1     Running   0          5m23s
local-path-storage   local-path-provisioner-57c5987fd4-sgvtl     1/1     Running   0          5m15s

❯ kubectl patch service example-service -p '{"spec":{"externalTrafficPolicy":"Local"}}'
service/example-service patched

❯ kubectl get pods -A                                                                  
NAMESPACE            NAME                                        READY   STATUS    RESTARTS     AGE
...
kube-system          kube-controller-manager-ovn-control-plane   0/1     Running   1 (2s ago)   6m25s
...
```

### Anything else we need to know?

_No response_

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: v1.30.2
Kustomize Version: v5.0.4-0.20230601165947-6ce0bf390ce3
Server Version: v1.31.0
```

</details>

### Cloud provider

<details>
None
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>

</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

### PR DESCRIPTION
**Pull request by carlory**
#### What type of PR is this?

/kind bug
/kind regression

#### What this PR does / why we need it:

If the init fails, we shouldn't register event handlers. It will cause a crash since https://github.com/kubernetes/kubernetes/pull/122145 is merged. 

There's a memory leak in the code. the queue has no worker to consume the event.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/128121

#### Does this PR introduce a user-facing change?

```release-note
Fixes 1.31 regression that can crash kube-controller-manager's service-lb-controller loop
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by carlory** [comment]
Test result:

```
(base) ➜  kubernetes git:(fix-128121-1) kind create cluster --name fix-128121-1 --image kindest/node:fix-128121-1
Creating cluster "fix-128121-1" ...
 ✓ Ensuring node image (kindest/node:fix-128121-1) 🖼
 ✓ Preparing nodes 📦
 ✓ Writing configuration 📜
 ✓ Starting control-plane 🕹️
 ✓ Installing CNI 🔌
 ✓ Installing StorageClass 💾
Set kubectl context to "kind-fix-128121-1"
You can now use your cluster with:

kubectl cluster-info --context kind-fix-128121-1

Have a nice day! 👋
(base) ➜  kubernetes git:(fix-128121-1) kubectl create -f __testdata/service.yaml
service/example-service created
(base) ➜  kubernetes git:(fix-128121-1) kubectl get pods -A
NAMESPACE            NAME                                                 READY   STATUS    RESTARTS   AGE
kube-system          coredns-7c65d6cfc9-4vjhj                             1/1     Running   0          2m46s
kube-system          coredns-7c65d6cfc9-plg98                             1/1     Running   0          2m46s
kube-system          etcd-fix-128121-1-control-plane                      1/1     Running   0          2m52s
kube-system          kindnet-7579v                                        1/1     Running   0          2m46s
kube-system          kube-apiserver-fix-128121-1-control-plane            1/1     Running   0          2m52s
kube-system          kube-controller-manager-fix-128121-1-control-plane   1/1     Running   0          2m52s
kube-system          kube-proxy-gsdn9                                     1/1     Running   0          2m46s
kube-system          kube-scheduler-fix-128121-1-control-plane            1/1     Running   0          2m52s
local-path-storage   local-path-provisioner-57c5987fd4-rpj4b              1/1     Running   0          2m46s
(base) ➜  kubernetes git:(fix-128121-1) kubectl patch service example-service -p '{"spec":{"externalTrafficPolicy":"Local"}}'
service/example-service patched
(base) ➜  kubernetes git:(fix-128121-1) kubectl get pods -A
NAMESPACE            NAME                                                 READY   STATUS    RESTARTS   AGE
kube-system          coredns-7c65d6cfc9-4vjhj                             1/1     Running   0          3m46s
kube-system          coredns-7c65d6cfc9-plg98                             1/1     Running   0          3m46s
kube-system          etcd-fix-128121-1-control-plane                      1/1     Running   0          3m52s
kube-system          kindnet-7579v                                        1/1     Running   0          3m46s
kube-system          kube-apiserver-fix-128121-1-control-plane            1/1     Running   0          3m52s
kube-system          kube-controller-manager-fix-128121-1-control-plane   1/1     Running   0          3m52s
kube-system          kube-proxy-gsdn9                                     1/1     Running   0          3m46s
kube-system          kube-scheduler-fix-128121-1-control-plane            1/1     Running   0          3m52s
local-path-storage   local-path-provisioner-57c5987fd4-rpj4b              1/1     Running   0          3m46s
```

File: staging/src/k8s.io/cloud-provider/controllers/service/controller.go
Comment (aojea) [inline staging/src/k8s.io/cloud-provider/controllers/service/controller.go:130]: yeah, do we have the same bug in the other controllers?

Comment (carlory) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:130]: other cloud controllers which are used by kcm have checked it in `New` func

Comment (carlory) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:130]: - https://github.com/carlory/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/nodelifecycle/node_lifecycle_controller.go#L81

Comment (carlory) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:130]: https://github.com/carlory/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go#L116 (not used by kcm)

Comment (carlory) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:130]: https://github.com/carlory/kubernetes/blob/master/staging/src/k8s.io/cloud-provider/controllers/route/route_controller.go#L72 doesn't have a cloud field.

Comment (aojea) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:130]: yes, we do, it turns out some controllers do check if there is a cloud provider not nil on the start functions, see

https://github.com/kubernetes/kubernetes/blob/6cc99274872264a231c61ea373fb0bed1d9e2331/cmd/kube-controller-manager/app/core.go#L129-L133

https://github.com/kubernetes/kubernetes/blob/6cc99274872264a231c61ea373fb0bed1d9e2331/cmd/kube-controller-manager/app/core.go#L297-L300

In addition to this change, let's move those checks to that place too

```
iff --git a/cmd/kube-controller-manager/app/core.go b/cmd/kube-controller-manager/app/core.go
index b3637f273af..164323009c9 100644
--- a/cmd/kube-controller-manager/app/core.go
+++ b/cmd/kube-controller-manager/app/core.go
@@ -92,6 +92,10 @@ func newServiceLBControllerDescriptor() *ControllerDescriptor {
 }
 
 func startServiceLBController(ctx context.Context, controllerContext ControllerContext, controllerName string) (controller.Interface, bool, error) {
+       if controllerContext.Cloud == nil {
+               logger.Info("Warning: node-controller is set, but no cloud provider specified. Will not configure node lifecyle controller.")
+               return nil, false, nil
+       }
        serviceController, err := servicecontroller.New(
                controllerContext.Cloud,
                controllerContext.ClientBuilder.ClientOrDie("service-controller"),
@@ -261,6 +265,10 @@ func newCloudNodeLifecycleControllerDescriptor() *ControllerDescriptor {
 
 func startCloudNodeLifecycleController(ctx context.Context, controllerContext ControllerContext, controllerName string) (controller.Interface, bool, error) {
        logger := klog.FromContext(ctx)
+       if controllerContext.Cloud == nil {
+               logger.Info("Warning: node-controller is set, but no cloud provider specified. Will not configure node lifecyle controller.")
+               return nil, false, nil
+       }
        cloudNodeLifecycleController, err := cloudnodelifecyclecontroller.NewCloudNodeLifecycleController(
                controllerContext.InformerFactory.Core().V1().Nodes(),
                // cloud node lifecycle controller uses existing cluster role from node-controller
```

Comment (carlory) [reply staging/src/k8s.io/cloud-provider/controllers/service/controller.go:130]: added.

**Comment by aojea** [comment]
can we have an unit test @carlory ? I think `cmd/kube-controller-manager/app/controllermanager_test.go` may have somethign we can use to check those controllers are not started if Cloud is nil

**Comment by aojea** [comment]
the job integration failure seems related, wonder if we have some code depending on this behavior 🤔

**Comment by carlory** [comment]
> I think cmd/kube-controller-manager/app/controllermanager_test.go may have somethign we can use to check those controllers are not started if Cloud is nil

Need to add a new unit test. I will do it later.

**Comment by carlory** [comment]
run it again. If it still fails, I will have a look
/test pull-kubernetes-integration

**Comment by aojea** [comment]
one of these node controllers seesm required
```
  I1018 11:11:51.935654 70147 wait.go:274] Unschedulable nodes= 3, maximum value for starting tests= 0
  I1018 11:11:51.935722 70147 wait.go:276] 	-> Node kind-control-plane [[[ Ready=true, Network(available)=false, Taints=[{node-role.kubernetes.io/control-plane  NoSchedule <nil>} {node.kubernetes.io/not-ready  NoSchedule <nil>}], NonblockingTaints=node-role.kubernetes.io/control-plane ]]]
  I1018 11:11:51.935737 70147 wait.go:276] 	-> Node kind-worker [[[ Ready=true, Network(available)=false, Taints=[{node.kubernetes.io/not-ready  NoSchedule <nil>}], NonblockingTaints=node-role.kubernetes.io/control-plane ]]]
  I1018 11:11:51.935753 70147 wait.go:276] 	-> Node kind-worker2 [[[ Ready=true, Network(available)=false, Taints=[{node.kubernetes.io/not-ready  NoSchedule <nil>}], NonblockingTaints=node-role.kubernetes.io/control-plane ]]]
```

**Comment by aojea** [comment]
one of these node controllers seesm required
```
  I1018 11:11:51.935654 70147 wait.go:274] Unschedulable nodes= 3, maximum value for starting tests= 0
  I1018 11:11:51.935722 70147 wait.go:276] 	-> Node kind-control-plane [[[ Ready=true, Network(available)=false, Taints=[{node-role.kubernetes.io/control-plane  NoSchedule <nil>} {node.kubernetes.io/not-ready  NoSchedule <nil>}], NonblockingTaints=node-role.kubernetes.io/control-plane ]]]
  I1018 11:11:51.935737 70147 wait.go:276] 	-> Node kind-worker [[[ Ready=true, Network(available)=false, Taints=[{node.kubernetes.io/not-ready  NoSchedule <nil>}], NonblockingTaints=node-role.kubernetes.io/control-plane ]]]
  I1018 11:11:51.935753 70147 wait.go:276] 	-> Node kind-worker2 [[[ Ready=true, Network(available)=false, Taints=[{node.kubernetes.io/not-ready  NoSchedule <nil>}], NonblockingTaints=node-role.kubernetes.io/control-plane ]]]
```

File: cmd/kube-controller-manager/app/core.go
Comment (aojea) [inline cmd/kube-controller-manager/app/core.go:215]: this does not need cloud and this is the one breaking, you confused this with `newCloudNodeLifecycleControllerDescriptor`
 You can identify the controllers that need cloud not nil via the initFunc in the constructor that has `isCloudProviderController`

```

func newServiceLBControllerDescriptor() *ControllerDescriptor {
	return &ControllerDescriptor{
		name:                      cpnames.ServiceLBController,
		aliases:                   []string{"service"},
		initFunc:                  startServiceLBController,
		isCloudProviderController: true,
	}
}
```

Comment (aojea) [reply cmd/kube-controller-manager/app/core.go:215]: the one you have to change is just below
```

func newCloudNodeLifecycleControllerDescriptor() *ControllerDescriptor {
	return &ControllerDescriptor{
		name:                      cpnames.CloudNodeLifecycleController,
		aliases:                   []string{"cloud-node-lifecycle"},
		initFunc:                  startCloudNodeLifecycleController,
		isCloudProviderController: true,
	}
}
func startCloudNodeLifecycleController(ctx context.Context, controllerContext ControllerContext, controllerName string) (controller.Interface, bool, error) {
	logger := klog.FromContext(ctx)
```

**Comment by aojea** [comment]
/lgtm
/approve

This fix is targeted so we can backport it and solve the regression, once we merge this we can remove these controllers entirely with https://github.com/kubernetes/kubernetes/pull/128197

/assign @liggitt @dims @elmiko @cheftako

**Comment by carlory** [comment]
test result:
```
(base) ➜  kubernetes git:(fix-128121-1) kubectl create -f __testdata/service.yaml
service/example-service created
(base) ➜  kubernetes git:(fix-128121-1) kubectl get pods -A
NAMESPACE            NAME                                                 READY   STATUS    RESTARTS   AGE
kube-system          coredns-7c65d6cfc9-mhcbk                             0/1     Pending   0          9s
kube-system          coredns-7c65d6cfc9-rwl88                             0/1     Pending   0          9s
kube-system          etcd-fix-128121-1-control-plane                      1/1     Running   0          15s
kube-system          kindnet-h9b95                                        1/1     Running   0          9s
kube-system          kube-apiserver-fix-128121-1-control-plane            1/1     Running   0          15s
kube-system          kube-controller-manager-fix-128121-1-control-plane   1/1     Running   0          15s
kube-system          kube-proxy-4dlz5                                     1/1     Running   0          9s
kube-system          kube-scheduler-fix-128121-1-control-plane            1/1     Running   0          15s
local-path-storage   local-path-provisioner-57c5987fd4-vkx92              0/1     Pending   0          9s
(base) ➜  kubernetes git:(fix-128121-1) kubectl patch service example-service -p '{"spec":{"externalTrafficPolicy":"Local"}}'
service/example-service patched
(base) ➜  kubernetes git:(fix-128121-1) kubectl get pods -A
NAMESPACE            NAME                                                 READY   STATUS    RESTARTS   AGE
kube-system          coredns-7c65d6cfc9-mhcbk                             1/1     Running   0          100s
kube-system          coredns-7c65d6cfc9-rwl88                             1/1     Running   0          100s
kube-system          etcd-fix-128121-1-control-plane                      1/1     Running   0          106s
kube-system          kindnet-h9b95                                        1/1     Running   0          100s
kube-system          kube-apiserver-fix-128121-1-control-plane            1/1     Running   0          106s
kube-system          kube-controller-manager-fix-128121-1-control-plane   1/1     Running   0          106s
kube-system          kube-proxy-4dlz5                                     1/1     Running   0          100s
kube-system          kube-scheduler-fix-128121-1-control-plane            1/1     Running   0          106s
local-path-storage   local-path-provisioner-57c5987fd4-vkx92              1/1     Running   0          100s
(base) ➜  kubernetes git:(fix-128121-1) kubectl -n kube-system logs kube-controller-manager-fix-128121-1-control-plane | grep "Warning: skipping controller"
I1019 08:01:18.997028       1 controllermanager.go:775] "Warning: skipping controller" controller="cloud-node-lifecycle-controller"
I1019 08:01:19.031232       1 controllermanager.go:775] "Warning: skipping controller" controller="service-lb-controller"
I1019 08:01:19.820578       1 controllermanager.go:775] "Warning: skipping controller" controller="storage-version-migrator-controller"
I1019 08:01:23.275656       1 controllermanager.go:775] "Warning: skipping controller" controller="node-route-controller"
```

**Review by soltysh — APPROVED** [review]
The fix lgtm, but the added unit test doesn't prove it, it will pass with and w/o the fix, so I'd probably want to see a test added which will ensure this is not broken by accident.

/approve

/hold 
for the additional test

File: cmd/kube-controller-manager/app/controllermanager_test.go
Comment (aojea) [inline cmd/kube-controller-manager/app/controllermanager_test.go:232]:
```suggestion
	controllerCtx.ComponentConfig.Generic.Controllers = []string{"*"}
```

This requires to enable all controllers to run, otherwise it does not try to run

@soltysh with this change it fails without the patch

Comment (soltysh) [reply cmd/kube-controller-manager/app/controllermanager_test.go:232]: Perfect, let's get that added - thank you!

**Comment by aojea** [comment]
/lgtm

@thockin can you approve the service controller changes?

**Comment by dims** [comment]
cc @oliviassss

**Pull request merged**
